#455475 [C#テキスト]2章演習問題

kurosu-being opened the pull request:
第2章 C#でプログラムを書いてみよう

1. メソッドの作り方、扱い方を学び、距離換算プログラムを作成する。
2. CSVファイルの読み込み、売上集計プログラムを作成する
3. クラスをインターフェースに書き換える
4. varによる暗黙の型指定
などを学習しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

第2章演習問題の新規プロジェクトを作成しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

演習問題２の問１-1を書きこみ、1-2の途中までを書き込みました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-1,問1-2を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-3までを実装し、問1-4のファイルを作成しました、

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-4を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-4においてコーディング規約に則りメソッドにコメントを追加しました

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問2を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問３の途中まで（クラスを書く）を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問3を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問2,問3の問題文を変更、追加しました。
問1のローカル変数song1,song2,song3をコーディング規約に則り変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1において不要なusingを削除しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

foreach文の中のローカル変数の命名を変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

インターフェイスの変数の名前をISale → Saleに変更しました、

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Saleクラス名を変更し、フィールドのSaleをFSakesに変更しました。

k-miyahara-being commented on 章末問題2/Practice2-1-1/Song.cs:
```diff
@@ -0,0 +1,16 @@
+﻿namespace Practice2_1_1 {
+    class Song {
```
前回指摘が漏れていたかもしれません。
自作したクラスにも何のクラスかわかるようにドキュメントコメント(ヘッダーコメント)を付けてください。
Programクラスには不要です。

kurosu-being replied:
ご指摘ありがとうございます！
Program以外のクラスにドキュメントコメントを付けるように修正いたしました。

k-miyahara-being commented on 章末問題2/Practice2-1-3/Program2-1-3.cs:
```diff
@@ -0,0 +1,12 @@
+            Song wSong2 = new Song("アイドル", "YOASOBI", 210);
+            Song wSong3 = new Song("天体観測", "BUMP OF CHICKEN", 190);
+            //配列songsに格納
+            Song[] songs = { wSong1, wSong2, wSong3 };
```
`songs`がコーディング規約違反です。
惜しい！

私が研修を受けたときも「白日」を書いた気がする...😂

kurosu-being replied:
ご指摘ありがとうございます！songs → wSongsに修正いたしました。

タイトルが短い曲、で一番最初に思いついたのが白日でした^^

k-miyahara-being commented on 章末問題2/Practice2-1-4/Calc.cs:
```diff
@@ -0,0 +1,26 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
不要なusingが残っています。

kurosu-being replied:
ご指摘ありがとうございます。不要なusingを削除いたしました。

k-miyahara-being commented on 章末問題2/Practice2-1-4/Calc.cs:
```diff
@@ -0,0 +1,26 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Practice2_1_4 {
+    class Calc {
+        /// <summary>
```
時間計算用のクラスを作ったのは良い判断だと思います！
クラス名の命名についても少し意識してみましょう。

そもそも`Calc`のような略語はあまり使わないことをおすすめします。
人によって意味が通じないことがあるためです。
単語が長いから短くする、という場合もありますが、長い単語 = 一般的に使われない ともいえるので、別の単語に置き換えられないか検討するのが良いでしょう。

計算するクラスなら、Calcul時間計算用のクラスを作ったのは良い判断だと思います！
クラス名の命名についても少し意識してみましょう。

そもそも`Calc`のような略語はあまり使わないことをおすすめします。
人によって意味が通じないことがあるためです。
単語が長いから短くする、という場合もありますが、長い単語 = 一般的に使われない ともいえるので、別の単語に置き換えられないか検討するのが良いでしょう。

計算するためのクラスなら、`Calculator`などがいいんじゃないでしょうか。

kurosu-being replied:
ご指摘ありがとうございます！クラスを作ったまでは良いものの格好つけて短くしたのが仇でした🙃誰でも、すぐに分かる命名が大切なのですね。名前が長くなってしまうときは上手く言い換えられるよう検索などを活用したいと思います。

k-miyahara-being commented on 章末問題2/Practice2-1-4/Calc.cs:
```diff
@@ -0,0 +1,26 @@
+namespace Practice2_1_4 {
+    class Calc {
+        /// <summary>
+        /// 分を計算するメソッド
+        /// </summary>
+        /// <param name="vLength">曲の長さ</param>
+        /// <returns>曲の分/returns>
+        public static int MinuteLength(int vLength){
+            return vLength / 60;
+        }
```
メソッド名には、基本的に動詞が含まれていることがふさわしいです。（メソッドは「～する」ものだから）
そのメソッドで行うことを、端的に動詞（または動詞+名詞）で置き換えてみてください。

kurosu-being replied:
ご指摘ありがとうございます！CalculateMinute、CalculateSecondに命名しなおしましたが、長いでしょうか。。。

k-miyahara-being replied:
長さは問題ないと思います。
しかし、単語を見直してみても良いかもしれません。
確かに中の処理としては「計算」していますが、このメソッドの役割は、分を秒(秒を分)に「変換」することです。
Calculateという単語はかなり汎用的で、使う場合は何をどのように計算するのかが明確になっていないと、誤解を生みかねません。

k-miyahara-being reviewed (changes requested):
とりあえず見れたところまで。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Program以外のクラスにドキュメントコメントを付けるように修正しました。
songsなどローカル変数の命名を変更しました。
不要なusingを削除いたしました。
クラスに対しヘッドーコメントを追加しました。
メソッドの名前を動詞＋名詞の端的なものに変更しました。また、一部文言を変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

不要な空行を削除しました。

ko-kato-being reviewed (commented):
「Practice1-2-4」は必要な差分ですか？
2章のプルリクエストなので、要らないんじゃないかなと思って聞いてみました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

不要なファイルを削除しました。

k-miyahara-being commented on 章末問題2/Practice2-1-4/Program2-1-4.cs:
```diff
@@ -0,0 +1,20 @@
+    class Program {
+        //Practice2-1-4 配列に格納されたすべてのSongオブジェクトの内容をコンソールに出力してください。演奏時間の表示は「4:16」のような書式にしてください。ただし、演奏時間は必ず60分未満と仮定してかまいません。
+        static void Main(string[] args) {
+            Song wSong1 = new Song("白日", "KingGnu", 241);
+            Song wSong2 = new Song("アイドル", "YOASOBI", 275);
+            Song wSong3 = new Song("天体観測", "BUMP OF CHICKEN", 1190);
+            //配列songsに格納
+            Song[] wSongs = { wSong1, wSong2, wSong3 };
```
[追加問題]
11行目のあと、さらに新しくSongクラスのインスタンスを生成して、wSongsに追加する処理を書いてみてください。

kurosu-being replied:
ありがとうございます！
Practice2-1-4(追加問題)というファイルを作成し、wSongsに君が代を追加しました。

k-miyahara-being replied:
要素数がかわりやすいものに対しては、配列よりもリストを使うのが良いです。
リストを宣言して書き直してみてください。

kurosu-being replied:
リストに宣言して書き直してみました！ご確認よろしくお願いします。

k-miyahara-being replied:
すみません、伝え方が悪かったです。
新しくListのインスタンスをつくるのではなく、そもそもwSongsをList型にしておけばよいということです。

kurosu-being replied:
ご指摘の意図を取り違えていました。申し訳ないです🙇‍♀️
wSongsをList型に変更いたしました。

k-miyahara-being replied:
修正不要ですが、変数名はwSongsのままで良かったですね

kurosu-being replied:
コメントも配列ｗSongsになっていたので、wSongsで修正いたしました！

k-miyahara-being commented on 章末問題2/Practice2-1-4/Program2-1-4.cs:
```diff
@@ -0,0 +1,20 @@
+            Song[] wSongs = { wSong1, wSong2, wSong3 };
+            // 配列に格納されたSongクラスのインスタンスをコンソールに出力
+            foreach (Song wSong in wSongs) {
+                int wShowMinute = Calculator.CalculateMinute(wSong.Length);
+                int wShowSecond = Calculator.CalculateSecond(wSong.Length);
+                Console.WriteLine($"{wSong.Title}という曲のアーティストは {wSong.ArtistName}で曲の長さは{wShowMinute:D1}:{wShowSecond:D2}です");
```
[追加問題]
自作メソッドを使わずに、[TimeSpan構造体](https://learn.microsoft.com/ja-jp/dotnet/api/system.timespan?view=net-7.0)を使って同等の処理を書いてみてください。
※前のコードは残したままでよいです

kurosu-being replied:
Practice2-1-4(追加問題)において、TimeSpan構造体を用いて同等の処理を書きました。
ご確認よろしくお願いいたします。

k-miyahara-being commented on 章末問題2/Practice2-2-1/ConvertFeet.cs:
```diff
@@ -0,0 +1,17 @@
+    /// メートルからインチ変換するクラス
+    /// </summary>
+    public static class ConvertInch {
+        private const double C_Ratio = 0.0254;
```
定義された定数がどこにも使われていないです。

kurosu-being replied:
ご指摘ありがとうございます。
0.0254を使用した計算式に`C_Ratio`を使用して変更しました。
（一度しか使わない定数でしたが、一般的な定数なので後からメソッドの追加や変更のしやすいように
定数として宣言しました）

k-miyahara-being replied:
> 一度しか使わない定数でしたが、一般的な定数なので後からメソッドの追加や変更のしやすいように
定数として宣言しました

Goodです！
基本的に、ロジックの中にリテラル（定数のこと）を直接書くことは好ましくありません。
例えば消費税率など、変化する可能性のある値を複数箇所で直接書いていると、
いざその値を変更しなければならないとなったときに、規模が大きなシステムほど修正が大変です。
定数として宣言して利用しておけば、その1か所を変更するだけで対応が完了します。
さらに、定数定義用のstaticなクラスを用意してまとめて宣言しておくと、より使いやすくメンテナンスしやすくなります。
ただし、特定の1か所でしか使われないような値は、そのまま書いてしまう、でもよいと思います。

kurosu-being replied:
ありがとうございます😃

> さらに、定数定義用のstaticなクラスを用意してまとめて宣言しておくと、より使いやすくメンテナンスしやすくなります。

勉強になります。確かにその方法が一番シンプルかつ分かりやすいですね。変数を新たに定義すべきか、そのまま書くべきか後のことも考えた視点を持ちたいと思います。

k-miyahara-being commented on 章末問題2/Practice2-2-1/ConvertFeet.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// メートルからインチ変換するクラス
+    /// </summary>
+    public static class ConvertInch {
```
ここのクラス名も、動詞よりも名詞がふさわしいです。
InchConverterとかですかね！

kurosu-being replied:
ご指摘ありがとうございます！
クラス名を`InchConverter`に変更し、問3においても`CountSales`→`SalesCounter`などに名称変更しました。
クラス名も動詞＋名詞にすると誤解していました。（メソッドなど関数を分かりやすくするための決まりごとですね）

k-miyahara-being commented on 章末問題2/Practice2-2-1/Program2-2-1.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <param name="vStop">終わりの数値</param>
+        static void PrintMeterToInchList(int vStart, int vStop) {
+            for (int wMeter = vStart; wMeter <= vStop; wMeter++) {
+                double wInch = ConvertInch.FromMeter(wMeter);
+                Console.WriteLine($"{wMeter}m = {wInch:0.0000}in");
```
出力結果がまちがっているので確認してください。
問題は、「インチからメートルへ変換する」とのことなので、使うべきメソッドが違うかなと思います。

kurosu-being replied:
ご指摘ありがとうございます。
インチからメートルへに換算して表示するようにメソッドを変更しました。
問題を途中からインチとメートルを逆に誤認しており、コメントも変なことになっていました😇

k-miyahara-being commented on 章末問題2/Practice2-3-1/Program2-3-1.cs:
```diff
@@ -0,0 +1,14 @@
+        //「2.2:売上集計プログラム」で作成したプログラムを変更し、商品カテゴリ別の売上高を求めるプログラムを作成してください
+        static void Main(string[] args) {
+            var wSales = new CountSales("sales.csv");
+            var wAmountPerItem = wSales.GetPerItemSales();
+            foreach (var wObj in wAmountPerItem) {
```
`wAmountPerItem`はわざわざ宣言せず、右辺を直接9行目に書いてしまってよいです。

kurosu-being replied:
ご指摘ありがとうございます！
`wAmountPerItem`を削除し、foreach文の中で直接`wSales.GetPerItemSales()`を使用するように変更しました。
一度しか使わない変数はわざわざ宣言しなくてもよいですね。

k-miyahara-being commented on 章末問題2/Practice2-3-1/Program2-3-1.cs:
```diff
@@ -0,0 +1,14 @@
+    class Program {
+        //「2.2:売上集計プログラム」で作成したプログラムを変更し、商品カテゴリ別の売上高を求めるプログラムを作成してください
+        static void Main(string[] args) {
+            var wSales = new CountSales("sales.csv");
```
実行すると、`sales.csv`というファイルがみつからずエラーが起きます。
必要なファイルを適切な位置に配置してください。

kurosu-being replied:
ご指摘ありがとうございます！
ファイルの位置を変更し、相対パスで参照するように変更いたしました。
ご確認よろしくお願い致します

k-miyahara-being reviewed (changes requested):
かなり指摘が少なく、おおかたOKです！
追加問題を出してみたので解いてみてください

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問2においてメソッドの名前をCalculateMinute→ConvertMinuteに変更しました。同様にCalculateSecond→ConvertSecondに変更しました。
問2において、問題文及びメソッドの中身をインチからメートルに返還するよう修正しました。
問2、問3においてクラスの名称を名詞に変更しました。
問3においてsales.csvファイルの配置を変更し、相対パスで参照するよう変更しました。
問3においてwAmountPerItemという変数を削除し、foreachの中で直接wSales.GetPerItemSalesを呼ぶように変更しました。

kurosu-being commented:
> 「Practice1-2-4」は必要な差分ですか？ 2章のプルリクエストなので、要らないんじゃないかなと思って聞いてみました。

ご指摘ありがとうございます！不要なファイルでしたので、削除いたしました。

ko-kato-being reviewed (commented):
@kurosu-being 
>ご指摘ありがとうございます！不要なファイルでしたので、削除いたしました。

修正ありがとうございました。
コミットする際に、必要な差分かどうか確認する癖をつけていけると良いですねー。
とはいえ、見落としはどうしてもあると思います。
Github の Files changed タブで、マージされる差分をすべて確認できるので、最終的にはここも確認すると良いかもしれません。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

追加問題を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

追加問題の問題文を追加しました。

k-miyahara-being commented on 章末問題2/Practice2-3-1/CountSales.cs:
```diff
@@ -0,0 +1,51 @@
+        /// <returns>Saleオブジェクトのリスト</returns>
+        private static IEnumerable<Sale> ReadSales(string vFilePath) {
+            var wSales = new List<Sale>();
+            var wLines = File.ReadLines(vFilePath);
+            foreach (var wLine in wLines) {
```
`wLines`も不要な変数ですね。
また、foreach文の`var`を明示的な型にしておきましょう。
`var`は、第三者がその行を見て、変数の型が容易に推測できる場合に使用するとよいです。
例えば、インスタンス生成などで右辺をみれば型が書いてある場合などです。(23行目のような)

kurosu-being replied:
ご指摘ありがとうございます。不要な変数を削除し、foreach文のvarをstringに明示しました。

k-miyahara-being replied:
varの使いどきで、変数の型があきらかってどういうこと？と思ったら、Visual Studioで下記の設定をしておくことをおすすめします。
ツール > オプション > テキストエディタ― > C# > コードスタイル > 全般
「'var'を優先 - 変数の型が明らかな場合」を「'var'を優先してください」「提案事項」
![image](https://github.com/kurosu-being/TextC/assets/86827185/439383ae-f629-4a2c-8055-68e61b546f80)

こうすると、ソースコードに「...」がついて直してみてはどう？と提案してくれるようになります。

![image](https://github.com/kurosu-being/TextC/assets/86827185/b6088ef8-1092-464b-a6e4-7296dc893f1d)

kurosu-being replied:
便利な機能ですね！紹介してくださりありがとうございます😃
Visual Studioそのものについて、どんな機能があるのか、まだ知らないことが多いので
個人的にも調べる時間をどこかでとりたいと思います。

k-miyahara-being commented on 章末問題2/Practice2-3-1/Sale.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// 売上情報のクラス
+    /// </summary>
+    class Sale {
+        /// <summary>
+        /// 店舗名
```
コンストラクタを定義して、引数を使ってプロパティを設定しましょう。

今はインスタンスを生成するときに初期化子を使用しているので問題ありませんが、
第三者が誤って初期化子を使用しなかったとき、店舗名もカテゴリも売上高も持たないインスタンスが生成されてしまいます。
コンストラクタがあることで、強制的にプロパティに値をセットさせることができるので、不完全なインスタンスが生成させてしまうことを防ぐことができます。

kurosu-being replied:
ご指摘ありがとうございます！引数を使ってプロパティの設定を行いました。
不完全なインスタンスを生成しない為にも、プロパティを実装したらコンストラクタもセットで実装した方が良いですね。

k-miyahara-being replied:
第3引数の接頭辞がまちがってますー！
また、作ってもらったコンストラクタの閉じかっこのインデント(字下げ)がずれています。

kurosu-being replied:
確認不足でした、申し訳ありません。
wAmount→vAmountに修正し、コード整形を行いました。

k-miyahara-being replied:
vAmoutになってますね😅

kurosu-being replied:
失礼しました😇修正いたしました。

k-miyahara-being commented on 章末問題2/Practice2-3-1/CountSales.cs:
```diff
@@ -0,0 +1,51 @@
+    /// 売上計算のクラス
+    /// </summary>
+    public class SalesCounter {
+        private IEnumerable<Sale> FSales;
```
フィールド変数FSalesは、クラス内で自由に書き替えできるようになっています。
ファイルから読み込んだデータは、書き換えされないことが安全です。
readonlyを付けることで読み取り専用の変数になり、コンストラクタ以外からの書き込みを防ぐことができます。

kurosu-being replied:
ご指摘ありがとうございます！
readonlyをつけて読み取り専用の変数に変更しました。
readonlyが初見でしたので、ファイル読み込みや外部データの読み込み等で使っていけるように学習したいと思います。

k-miyahara-being commented on 章末問題2/Practice2-3-1/CountSales.cs:
```diff
@@ -0,0 +1,51 @@
+        /// 商品カテゴリ別売上を求めるメソッド
+        /// </summary>
+        /// <returns>商品カテゴリ別売上</returns>
+        public IDictionary<string, int> GetPerItemSales() {
+            var wDict = new Dictionary<string, int>();
+            foreach (var wSale in FSales) {
+                if (wDict.ContainsKey(wSale.ProductCategory))
+                    wDict[wSale.ProductCategory] += wSale.Amount;
+                else
+                    wDict[wSale.ProductCategory] = wSale.Amount;
+            }
+            return wDict;
+        }
```
[追加問題]
3章で学習したLINQを使って書き直してみましょう。
同等の処理を最短1行で書けてしまいます。
ヒント：GroupBy, Sum

kurosu-being replied:
ありがとうございます！追加問題を実装しました。
GroupBy, Sumの扱い方を理解するのに時間がかかり、その過程で自分がラムダ式についてまだ理解が浅いことに気づきました。
ご確認よろしくお願いします。

k-miyahara-being replied:
OKです！
ちなみに、メソッドの中身がreturnしかないものは↓のように`{ }`なしで表現できます。
①returnにカーソルをあてる、②電球マーク→メソッドに式本体を使用する
![image](https://github.com/kurosu-being/TextC/assets/86827185/5b3b7f51-b352-4eb5-80c7-cb3f7a406d09)

この変形後の形、何かに似ていませんか？
これが何者かはまた調べてみてくださいー

kurosu-being replied:
変形を実行してみました！
メソッドそのものがラムダ式のような形になりますね😳
検索したらラムダ式本体によるメソッドの記述、と出てきました。ラムダ式のコード削減力が凄い。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

追加問題の配列をListに書き直しました。
[追加問題]3章で学習したLINQを使って書き直してみましょう。を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Practice2-3-1において、型を明示しました。

s-morita-being commented on 章末問題2/Practice2-3-1/Program2-3-1.cs:
```diff
@@ -0,0 +1,13 @@
+    class Program {
+        //「2.2:売上集計プログラム」で作成したプログラムを変更し、商品カテゴリ別の売上高を求めるプログラムを作成してください
+        static void Main(string[] args) {
+            var wSales = new SalesCounter(@"../../sales.csv");
```
相対パスの単純なハードコーディングは要注意です。

プロセス(プログラムの実行単位)は、それぞれ、自分が現在どのディレクトリにいるかという情報を管理しています。 (カレントディレクトリ current directory)

相対パスをそのまま指定すると、通常、**カレントディレクトリからの**相対パスとして扱われます。

カレントディレクトリは、プロセス起動時の指定で (プロセス自身でも) 変更できる情報なので、
相対パスがいつも同じファイルを指すとは限らなくなります。

ハードコーディングするにしても、例えば、プログラムファイルがある場所からの相対パスとみなして、
絶対パスを求めてから読み込むようにするとか。

まーでも、そもそもパスをハードコーディングせずに、プログラムの引数として受け取れるようにすれば、
もっと汎用的で有用なプログラムになりますね。

kurosu-being replied:
ご指摘ありがとうございます！
プログラムの引数として受け取れるように、最初のコンソール画面でユーザーが自身の保存したファイルのパスを入力する形に変更いたしました。
ハードコーディングという概念を初めて知りました。宮原さんに解説して頂き、
GaiaCloudでお客様にファイルを読み込んでもらうなど実際の状況と照らし合わして理解を深めることが出来ました。
```

ハードコーディングするにしても、例えば、プログラムファイルがある場所からの相対パスとみなして、
絶対パスを求めてから読み込むようにするとか。
```
こちらの方法についてはまだ方法が分からない為、相対パス・絶対パスについて学習をすすめたいと思います。

s-morita-being replied:
見慣れない言葉を使ってしまいました。
ハードコーディングは、簡単に言えば固定の情報を直接ソースコードに記述する行為ですが、
その弊害や実例は、検索したり ChatGPT 先生にきいてもらってもよいかと思います。
https://ja.wikipedia.org/wiki/%E3%83%8F%E3%83%BC%E3%83%89%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0

あと、修正してもらったコードの方にもコメントしましたが、プログラムの引数として受け取る方法もあります。

プログラムには、大きく分けて、対話型と、非対話型 (バッチ処理、一括処理) の 2 種類があります。

どういうプログラムにしたいかによりますが、
対話型ならユーザー入力、非対話型なら引数、で受け取るのが適当かと思います。

対話型は、とりあえず起動したら案内をみながら操作できるので、初心者に易しいですが、
繰り返し使う場合に、入力が面倒だったりします。
非対話型は、実行時に引数を指定する必要があるので、とっつきにくいですが、
バッチ処理などの他の処理と組み合わせられるので、応用範囲を広げやすいです。

s-morita-being reviewed (commented):
お疲れ様です。
ひとつコメントしました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Practice2-3-1において、プログラムの引数として受け取れるように、最初のコンソール画面でユーザーが自身の保存したファイルのパスを入力する形に変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

コンストラクタのドキュメントコメントを追加しました。

s-morita-being commented on 章末問題2/Practice2-3-1/Program2-3-1.cs:
```diff
@@ -0,0 +1,15 @@
+        //「2.2:売上集計プログラム」で作成したプログラムを変更し、商品カテゴリ別の売上高を求めるプログラムを作成してください
+        static void Main(string[] args) {
+            Console.WriteLine("sales.csvのパスを入力してください(sales.csvファイルはチケットに貼っています)");
+            string wInputText = Console.ReadLine();
```
今回はこれでよいですが、プログラムの引数 args で受け取れるようにすると、
色んな引数で実行できるようになって便利です。

kurosu-being replied:
ありがとうございます！Mainの引数を使う、ということを試したことがありませんでした。
また別のコードで書いてみたいと思います🙋‍♀️

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Practice2-1-4追加問題において、wSongをList型に変更しました。
Practice2-3-1において、メソッドの式本体を変形しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

vAmout→vAmountに変数名を変更いたしました

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

wList→wSongsに変更しました。
